### ISSUE TITLE
train for special characters

### ISSUE THREAD

**Issue by fsluckymao** [Canjie-Luo/MORAN_v2#44]
Hi
if I want to train ASRN to recognize Special characters like '#','@','%' etc, what should I do except expand alphabet?
thank your.

**Comment by Canjie-Luo** [comment]
Expanding the alphabet is enough. Note that we use "$" as EOS, you can replace it by anther symbol. Please use "UTF-8" for non-English characters.

**Comment by fsluckymao** [comment]
thank you.

**Comment by fsluckymao** [comment]
HI @Canjie-Luo 👍 :
If I want to train a model which can recognize character ,number and special character on my dataset, does it mean I need to train it from scratch rather than fine tuning your pre-train model(demo.pth)? 
Anothor question, how can I upload training dataset to server, they are too big = =\
thank you.

**Comment by Canjie-Luo** [comment]
You can fine-tune the pre-train model. Don't load the parameters of the attention module. Because your number of classes is larger than mine. The second question, emmm, I am not sure that what do you mean, do you want to put the dataset on your server? Try "scp"?

**Comment by absaravanan** [comment]
Hi.

I needed some help, i wanted to train with other charcters, i needed some help with removing the parameters of the attention module, 

i tried passing anything which starts with ASRN, like this 

    for k, v in state_dict.items():
            if k.startswith('module.ASRN') or k.startswith('ASRN'):
                continue
            else:
                name = k.replace("module.", "") # remove `module.`
                MORAN_state_dict_rename[name] = v

but this did not work , any suggestions would be helpful.Thanks

**Comment by DecentMakeover** [comment]
Hi.

I needed some help, i wanted to train with other charcters, i needed some help with removing the parameters of the attention module,

i tried passing anything which starts with ASRN, like this

for k, v in state_dict.items():
        if k.startswith('module.ASRN') or k.startswith('ASRN'):
            continue
        else:
            name = k.replace("module.", "") # remove `module.`
            MORAN_state_dict_rename[name] = v
but this did not work , any suggestions would be helpful.Thanks

**Comment by fsluckymao** [comment]
Hi:@Canjie-Luo @DecentMakeover 
I just i beginer,maybe you should set strict to 'False'
plz refer to 
[https://pytorch.org/tutorials/beginner/saving_loading_models.html#warmstarting-model-using-parameters-from-a-different-model](url)

**Comment by DecentMakeover** [comment]
@fsluckymao Thanks ,this works ,but my accuracy is zero ,not sure why.

Anyways,Thanks for the help !

**Comment by Canjie-Luo** [comment]
Thanks for both of your consideration. @fsluckymao @DecentMakeover

**Comment by DecentMakeover** [comment]
Hi 

@fsluckymao @Canjie-Luo , I just trained it on special characters overnight- 
This is what the alphabet look like - 

    alphabet = '0:1:2:3:4:5:6:7:8:9:a:b:c:d:e:f:g:h:i:j:k:l:m:n:o:p:q:r:s:t:u:v:w:x:y:z:/:.:~:!:@:#:%:^:&:*:(:):_:=:,:;:|:\:-:+:`:?:[:]:<:>:}:{:":":":":":$'

but given a test image it is not able to predict any of these characters.My test accuracy is around 88%.

Any ideas?

**Comment by fsluckymao** [comment]
Hi @DecentMakeover :
I am still doing experiment, I try to train MORAN with MORN and without MORN to make a Comparison.
because I haven't upload training dataset to my server,  I was used test_dataset(ic13_1015 and IIIT5K) and my own dataset(with special characters) for training. Adn used ic15_2077 for val.  The result isn't good.but i think it is because the lack of training dataset.
what is your training dataset for special characters and what test_dataset do you use for val?
good luck.

**Comment by DecentMakeover** [comment]
my data has sufficient special characters, but still it is not learning,even though the accuracy is 88%.
Not sure why its not learning, my question is, is alphabet the only change we have to make in the code?or is there any other modification that we have to do?

Thanks

**Comment by Canjie-Luo** [comment]
@DecentMakeover Please give the alphabet to the `lmdbdataset()` in `dataset.py`. Sorry, I forgot.

**Comment by Canjie-Luo** [comment]
@DecentMakeover Modify the line 68, 71, 82 in `main.py`: `lmdbDataset(...alphabet=opt.alphabet.split(opt.sep))`.

**Comment by absaravanan** [comment]
Okay, let me try this and check

**Comment by DecentMakeover** [comment]
Okay, let me try this and check.

**Comment by DecentMakeover** [comment]
Works.Amazing!

**Comment by pp3993** [comment]
Hi @Canjie-Luo , @DecentMakeover, what changes did you make to work it for special character ?
here is my list:
`'--alphabet', type=str, default='0:1:2:3:4:5:6:7:8:9:a:b:c:d:e:f:g:h:i:j:k:l:m:n:o:p:q:r:s:t:u:v:w:x:y:z:A:B:C:D:E:F:G:H:I:J:K:L:M:N:O:P:Q:R:S:T:U:V:W:X:Y:Z: :/:.:>:,:(:):-:$'`

Example datasets used for training. 

- ELGRENZVYULJI :2)\)7 ,91 < 9oR1#t4h80Njw

- Kgpyhzyszobme

- YBOBTOVJL  6<<'830 :)-4<9 .17

In my case training accuracy goes to 98% but still it is not learning. 
Also, as my data is having maximum width upto 565 pixels, Do i need to make any changes in targetW or ImageW parameter.

**Comment by Canjie-Luo** [comment]
What are the predictions in testing?

**Comment by Cabbagehust2507** [comment]
@Canjie-Luo I use your code to train with Vietnamese dataset but I'm not sure about some special character in Vietnamese language, I need to modify and write it like "\u0418:\u043b:\u00ab" or like "à:á:ả:ã:ạ:ă:ằ". Thank you so much! Hope that you can reply me soon! :)
